lib/check_elastic_search.rb

```diff
@@ -1,133 +1,144 @@
+    ms.id = Base64.encode64("#{self.class.name}/#{self.id}")
+    unless p.nil?
+      ms.team_id = p.team_id
+      ms.project_id = p.id
+    end
     rtid = self.is_a?(ProjectMedia) ? (self.related_to_id || self.sources.first&.id) : nil
     ms.relationship_sources = [Digest::MD5.hexdigest(Relationship.default_type.to_json) + '_' + rtid.to_s] unless rtid.blank?
     ms.set_es_annotated(self)
     self.add_extra_elasticsearch_data(ms)
+    ms.accounts = self.add_es_accounts if self.class.name == 'ProjectSource'
     ms.save!
   end
 
-  def update_media_search_bg(options)
-    ms = get_elasticsearch_parent(options[:parent])
-    unless ms.nil?
-      data = get_elasticsearch_data(options[:data])
-      # add mising field on parent
-      data.merge!(add_missing_fields(options))
-      fields = {'last_activity_at' => Time.now.utc}
-      options[:keys].each{|k| fields[k] = data[k] if ms.respond_to?("#{k}=") and !data[k].blank? }
-      ms.update fields
+  def add_es_accounts
+    ms_accounts = []
+    accounts = []
+    accounts = self.source.accounts unless self.source.nil?
+    accounts.each do |a|
+      ms_accounts << a.store_elasticsearch_data(%w(title description username), {})
     end
+    ms_accounts
   end
 
-  def add_missing_fields(options)
-    data = {}
-    parent = options[:parent]
-    return data unless ['ProjectMedia', 'ProjectSource'].include?(parent.class.name)
-    unless options[:keys].include?('project_id')
-      options[:keys] += ['team_id', 'project_id']
-      data.merge!({project_id: parent.project_id, team_id: parent.project.team_id})
-    end
-    if parent.class.name == 'ProjectMedia'
-      unless options[:keys].include?('status')
-        options[:keys] << 'status'
-        data.merge!({status: parent.last_status})
-      end
-      unless options[:keys].include?('title')
-        options[:keys] += ['title', 'description']
-        data.merge!({title: parent.title, description: parent.description})
-      end
-    end
-    data
+  def update_elasticsearch_doc(keys, data = {}, obj = nil)
+    return if self.disable_es_callbacks || RequestStore.store[:disable_es_callbacks]
+    options = {keys: keys, data: data}
+    options[:obj] = obj unless obj.nil?
+    ElasticSearchWorker.perform_in(1.second, YAML::dump(self), YAML::dump(options), 'update_doc')
+  end
+
+  def update_elasticsearch_doc_bg(options)
+    create_doc_if_not_exists(options)
+    sleep 1
+    data = get_elasticsearch_data(options[:data])
+    # data.merge!(add_missing_fields(options))
+    fields = {'updated_at' => Time.now.utc}
+    options[:keys].each{|k| fields[k] = data[k] if !data[k].blank? }
+    client = MediaSearch.gateway.client
+    client.update index: CheckElasticSearchModel.get_index_alias, type: 'media_search', id: options[:doc_id], retry_on_conflict: 3,
+                  body: { doc: fields }
   end
 
-  def add_update_media_search_child(child, keys, data = {}, parent = nil)
+  # Keep this method until verify search feature.
```

Comment:
what's that for?

Reply:
This method `add_missing_fields` was added to index all parent filed when using versions, now I use retry_on_conflict and just keep it until verify everything on QA then will remove it.
I did many tests locally and everything works well without this method.